Pre blocks

gertsonderby opened the pull request:
I added functions to create and remove PRE elements in a sensible way. I'm not fantastically happy with the removePreformatted function - I get the sense that there are easier, better ways to do this. But it has satisfying functionality at this juncture. Said functionality is also covered by some tests. These are not exhaustive, but ought to provide a good baseline for the functionality. I put the two functions under the key shortcut 'ctrl-shift-m', which will toggle based on hasFormat('pre') whether it creates or removes PREs.

(The things we do to make our systems engineers happy...)

gertsonderby commented:
Figured out a way to deal better with pasting into PRE tags. Will add commit shortly.

gertsonderby commented:
There we go. Pasting is much improved.

neilj commented on build/squire-raw.js:
```diff
@@ -1325,6 +1325,16 @@ var keyHandlers = {
             return;
         }
 
+        if ( /^PRE|CODE|SAMP$/.test( block.nodeName ) ) {
```
`CODE` and `SAMP` are both inline node types (see Node.js), so can never be returned as the block type for getStartBlockOfRange. This test should just look for `block.nodeName === 'PRE'`

neilj commented on build/squire-raw.js:
```diff
@@ -3386,6 +3405,89 @@ var decreaseListLevel = function ( frag ) {
+        ] );
+};
+
+var removePreformatted = function ( frag ) {
```
OK, this function has many code style violations; I'm not going to list them all. But beyond that, I'm not sure this is necessary at all. Can't you just do (not in a modifyBlocks call) something like:

    expandRangeToBlockBoundaries( range );
    moveRangeBoundariesUpTree( range, this._body );
    proto.removeAllFormatting();

Since your `<pre>` doesn't have any formatting in it anyway, all you really want to do is change it back to the default block type, which is exactly what removeAllFormatting will do.

gertsonderby replied:
I have to disagree that this echoes Squire#removeAllFormatting(). What I'm looking to do is maintain the line format as best it can be - and inside a PRE tag, whitespace linebreaks (`/\r?\n/`) are shown. So we'd want the content of the PRE element to be split by linebreaks, and turned into one defaultblock per line. As well, we need to split the PRE element at the range borders.

Both of those activities do have more graceful ways to do them than this, I'll grant. I'll make with finding them.

neilj commented on build/squire-raw.js:
```diff
@@ -3386,6 +3405,89 @@ var decreaseListLevel = function ( frag ) {
+    return this.createElement( 'PRE',
+        this._config.tagAttributes.pre, [
+            getTextFromHTMLFragment( this, frag )
+        ] );
```
No space between `]` and `)`.

gertsonderby commented:
Do you have any suggestions as to what removePreformatted() should do if collapsed inside a PRE element? As it stands, it does nothing, which I figured was reasonable.

gertsonderby commented:
Did a major clean-up pass. I think I got all the style errors, and I reworked removePreformatted() a fair bit. It will still split PRE blocks at range limits, but it refrains from using the weird range juggling I was doing. As well, it uses a TreeWalker to iterate through contained PRE blocks, instead of looping through a NodeList.

I also added a test (and a fix to make it work) to ensure that a range that fully encloses a PRE block and some other stuff acts right.

Does this pass muster?

gertsonderby commented:
Today I noticed I had forgotten to actually push stuff so it could be reviewed. GG, Gert. Well, I've pushed the changes I mentioned above! :-)

neilj commented:
Sorry for the delay in getting to this. So far testing this patch I have found the following issues:
* Selection/cursor position is not maintained when converting to preformatted.
* If my cursor is at the end of the pre block, I often have to hit enter *twice* to get a new line.
* If I unpreformat with my cursor somewhere in the `<pre>`, it just splits the pre rather than converting the whole line (or block) to normal.
* Some minor code formatting issues, where it isn't following the correct conventions; but I can fix this up myself if everything else is ready.

Unfortunately at this stage I can't merge this, sorry. There are too many issues.

gertsonderby commented:
Re. the third issue you mention, I think expanding selection to the line you're on makes sense. Inside a PRE block, you can think of the individual lines as blocks, so treating them separately for this purpose makes sense to me. I'm not a fan of expanding to the whole block - that seems like going a bit far, given the semantics in play.

I'll look into this. It's a not insignificant feature, so that it's taking a while to get right isn't fantastically surprising to me. :-)

gertsonderby commented:
A further issue found with this code: URLs written into PRE blocks get turned into links, which they really shouldn't.

gertsonderby commented:
Re. links in PREs, how *do* we feel about that? They're not hugely disruptive, per se, so it might be useful to keep them, at least the auto-inserted ones. But at the very least, I'll ensure that makeLink() behaves itself in there.

gertsonderby commented:
Anyhow, apart from the link business (which is very much a matter for discussion, I'd say), I think I've cleaned up the things you mentioned, @neilj. Let me know if you have other concerns with it. I hope I've cleaned up the messiness of it some.

neilj commented:
Just a quick note to say I haven't forgotten about this, but haven't managed to find enough time to review it properly yet, since it's quite a large patch!

gertsonderby commented:
I'll be doing a little maintenance work, so I'll also rebase this onto master so it's easier to work with.

gertsonderby commented:
I've got some time on my hands the next little while, so I figured it might be a good time to get this in order. Let me know what you think, @neilj.
